airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py

```diff
@@ -0,0 +1,273 @@
+
+    It supports multiple policies applied to different group of requests. To distinct these groups we use RequestMatchers.
+    Individual policy represented by CallRatePolicy and currently supports only moving window strategy.
+    """
+
+    @abc.abstractmethod
+    def add_policy(self, request_matcher: RequestMatcher, policy: CallRatePolicy) -> None:
+        """Add policy for calls
+
+        :param request_matcher: callable to match request object with corresponding policy
+        :param policy: to acquire calls
+        :return:
+        """
+
+    @abc.abstractmethod
+    def acquire_call(self, request: Any, block: bool = True, timeout: Optional[float] = None) -> None:
+        """Try to get a call from budget, will block by default
+
+        :param request:
+        :param block: when true (default) will block the current thread until call credit is available
+        :param timeout: if set will limit maximum time in block, otherwise will wait until credit is available
+        :raises: CallRateLimitHit - when no credits left and if timeout was set the waiting time exceed the timeout
+        """
+
+
+class APIBudget(AbstractAPIBudget):
+    """Default APIBudget implementation"""
+
+    def __init__(self) -> None:
+        """Constructor"""
+        self._policies: list[tuple[RequestMatcher, CallRatePolicy]] = []
+
+    def add_policy(self, request_matcher: RequestMatcher, policy: CallRatePolicy) -> None:
+        """Add policy for calls
+
+        :param request_matcher: callable to match request object with corresponding policy
+        :param policy: to acquire calls
+        :return:
+        """
+
+        self._policies.append((request_matcher, policy))
+
+    def acquire_call(self, request: Any, block: bool = True, timeout: Optional[float] = None) -> None:
+        """Try to get a call from budget, will block by default
+
+        :param request:
+        :param block: when true (default) will block the current thread until call credit is available
+        :param timeout: if set will limit maximum time in block, otherwise will wait until credit is available
+        :raises: CallRateLimitHit - when no credits left and if timeout was set the waiting time exceed the timeout
+        """
+
+        for matcher, policy in self._policies:
+            if matcher(request):
+                self._do_acquire(request, policy, block, timeout)
+                break
+
+    def _do_acquire(self, request: Any, policy: CallRatePolicy, block: bool, timeout: Optional[float]) -> None:
+        """Internal method to try to acquire a call credit
+
+        :param request:
+        :param policy:
+        :param block:
+        :param timeout:
+        """
+        try:
+            policy.try_acquire(request)
+        except CallRateLimitHit as exc:
+            if block:
+                if timeout is not None:
+                    time_to_wait = min(timedelta(seconds=timeout), exc.time_to_wait)
```

Comment:
`timeout` is the maximum time to wait, the logic and interfaces here resemble an interface of Queue. It will raise exception when timeout is less than time_to_wait and it intended behavior if caller provided us the timeout.

Reply:
But here, there is no indication that `timeout > exc.time_to_wait`, right? If this isn't the case, we shouldn't even wait, right?

Reply:
You are right that if we going to raise an exception because of `timeout < exc.time_to_wait` then there is no point in waiting, we can just raise an exception immediately. 
This was done to behave like blocking API, I don't have strong argument for it.